extensions/amp-a4a/0.1/real-time-config-manager.js

```diff
@@ -66,327 +66,351 @@ export const RTC_ERROR_ENUM = {
+   * @return {string}
+   */
+  getCalloutParam_(url) {
+    const parsedUrl = parseUrlDeprecated(url);
+    return (parsedUrl.hostname + parsedUrl.pathname).substr(0, 50);
   }
-  customMacros['TIMEOUT'] = () => rtcConfig['timeoutMillis'];
-  const promiseArray = [];
-  const seenUrls = {};
-  const rtcStartTime = Date.now();
-  // For each publisher defined URL, inflate the url using the macros,
-  // and send the RTC request.
-  (rtcConfig['urls'] || []).forEach(urlObj => {
-    let url, errorReportingUrl;
-    if (isObject(urlObj)) {
-      url = urlObj['url'];
-      errorReportingUrl = urlObj['errorReportingUrl'];
-    } else if (typeof urlObj == 'string') {
-      url = urlObj;
-    } else {
-      dev().warn(TAG, `Invalid url: ${urlObj}`);
+
+  /**
+   * For a given A4A Element, sends out Real Time Config requests to
+   * any urls or vendors specified by the publisher.
+   * @param {!Object<string, !../../../src/service/variable-source.AsyncResolverDef>} customMacros The ad-network specified macro
+   *   substitutions available to use.
+   * @param {?CONSENT_POLICY_STATE} consentState
+   * @return {Promise<!Array<!rtcResponseDef>>|undefined}
+   * @visibleForTesting
+   */
+  maybeExecuteRealTimeConfig(customMacros, consentState) {
+    // TODO(keithwrightbos) - allow pub to override such that some callouts are
+    // still allowed.
+    if (consentState == CONSENT_POLICY_STATE.INSUFFICIENT ||
+        consentState == CONSENT_POLICY_STATE.UNKNOWN) {
+      return;
     }
-    inflateAndSendRtc_(a4aElement, url, seenUrls, promiseArray,
-        rtcStartTime, customMacros,
-        rtcConfig['timeoutMillis'], errorReportingUrl);
-  });
-  // For each vendor the publisher has specified, inflate the vendor
-  // url if it exists, and send the RTC request.
-  Object.keys(rtcConfig['vendors'] || []).forEach(vendor => {
-    const vendorObject = RTC_VENDORS[vendor.toLowerCase()];
-    const url = vendorObject ? vendorObject.url : '';
-    const errorReportingUrl = vendorObject ?
-      vendorObject['errorReportingUrl'] : '';
-    if (!url) {
-      return promiseArray.push(
-          buildErrorResponse_(
-              RTC_ERROR_ENUM.UNKNOWN_VENDOR, vendor, errorReportingUrl,
-              a4aElement.win, a4aElement.getAmpDoc()));
+    this.validateRtcConfig_(this.a4aElement_.element);
+    if (!this.rtcConfig_) {
+      return;
     }
-    const validVendorMacros = {};
-    Object.keys(rtcConfig['vendors'][vendor]).forEach(macro => {
-      if (vendorObject.macros && vendorObject.macros.includes(macro)) {
-        const value = rtcConfig['vendors'][vendor][macro];
-        validVendorMacros[macro] = isObject(value) || isArray(value) ?
-          JSON.stringify(value) : value;
+    customMacros = this.assignMacros(customMacros);
+    this.rtcStartTime_ = Date.now();
+    this.handleRtcForCustomUrls(customMacros);
+    this.handleRtcForVendorUrls(customMacros);
+    return Promise.all(this.promiseArray_);
+  }
+
+  assignMacros(macros) {
```

Comment:
Could you add jsdoc to this method and the two below it?

Reply:
done